pkg/collector/corechecks/sbom/processor.go

```diff
@@ -210,49 +210,60 @@ func (p *processor) processContainerImagesRefresh(allImages []*workloadmeta.Cont
 	}
 }
 
-func (p *processor) processHostScanResult(result sbom.ScanResult) {
-	log.Debugf("processing host scanresult: %v", result)
+func (p *processor) processFilesystemScanResult(result sbom.ScanResult, kind, sbomID string, sbomType model.SBOMSourceType, errCallback func(err error) error) {
+	log.Debugf("processing %s scan result: %v", kind, result)
 	sbom := &model.SBOMEntity{
 		Status:             model.SBOMStatus_SUCCESS,
-		Type:               model.SBOMSourceType_HOST_FILE_SYSTEM,
-		Id:                 p.hostname,
+		Type:               sbomType,
+		Id:                 sbomID,
 		InUse:              true,
 		GeneratedAt:        timestamppb.New(result.CreatedAt),
 		GenerationDuration: convertDuration(result.Duration),
 	}
 
-	if result.Error != nil {
-		log.Errorf("Scan error: %v", result.Error)
-		sbom.Sbom = &model.SBOMEntity_Error{
-			Error: result.Error.Error(),
+	defer func() {
+		err := result.Error
+
+		if errCallback != nil {
+			if err = errCallback(err); err != nil {
+				log.Errorf("failed scan: %s", err)
+				return
+			}
 		}
-		sbom.Status = model.SBOMStatus_FAILED
-	} else {
-		log.Infof("Successfully generated SBOM for host: %v, %v", result.CreatedAt, result.Duration)
 
-		if p.hostCache != "" && p.hostCache == result.Report.ID() && result.CreatedAt.Sub(p.hostLastFullSBOM) < p.hostHeartbeatValidity {
-			sbom.Heartbeat = true
-		} else {
-			report, err := result.Report.ToCycloneDX()
-			if err != nil {
-				log.Errorf("Failed to extract SBOM from report: %s", err)
-				sbom.Sbom = &model.SBOMEntity_Error{
-					Error: err.Error(),
-				}
-				sbom.Status = model.SBOMStatus_FAILED
-			} else {
-				sbom.Sbom = &model.SBOMEntity_Cyclonedx{
-					Cyclonedx: convertBOM(report),
-				}
+		if result.Error != nil {
+			log.Errorf("Scan error: %v", result.Error)
+			sbom.Sbom = &model.SBOMEntity_Error{
+				Error: result.Error.Error(),
 			}
+			sbom.Status = model.SBOMStatus_FAILED
+		}
+
+		p.queue <- sbom
+	}()
 
-			sbom.Hash = result.Report.ID()
-			p.hostCache = result.Report.ID()
-			p.hostLastFullSBOM = result.CreatedAt
+	report, err := result.Report.ToCycloneDX()
+	if err != nil {
+		result.Error = err
+		return
+	}
+
+	sbom.Hash = string(result.Report.ID())
+
+	log.Infof("Successfully generated SBOM for %s: %v, %v", kind, result.CreatedAt, result.Duration)
+	if p.hostLastSBOMHash != "" && p.hostLastSBOMHash == sbom.Hash && result.CreatedAt.Sub(p.hostLastFullSBOM) < p.hostHeartbeatValidity {
```

Comment:
Do we want this host hash stuff in the procfs code path ?

Reply:
I would say yes